### PR TITLE
Integration of Apility.IO

### PR DESCRIPTION
**Pull request by devsuds**
new integration function - Apility.IO

Signed-off-by:
Sudip Sahoo

### CODE REVIEW

File: fn_apility/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+## About Apility.IO
+
+Apility.io can be defined as Threat Intelligence SaaS for developers and product companies that want to know in realtime if their existing or potential users have been classified as 'abusers' by one or more of these lists.
+
+Automatic extraction processes extracts all the information in realtime, keeping the most up to date data available, saving yourself the hassle of extract and update regularly all these lists and the data.
+
+## Using MxToolBox Function
```
Comment (mscherfling) [inline fn_apility/README.md]:
```suggestion
## Using the Ability.IO Function
```

File: fn_apility/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+## About Apility.IO
+
+Apility.io can be defined as Threat Intelligence SaaS for developers and product companies that want to know in realtime if their existing or potential users have been classified as 'abusers' by one or more of these lists.
+
+Automatic extraction processes extracts all the information in realtime, keeping the most up to date data available, saving yourself the hassle of extract and update regularly all these lists and the data.
```
Comment (mscherfling) [inline fn_apility/README.md]: I'm not sure sure what information this sentence is meant to convey. Maybe you can identify the different functions which this integration implements.

File: fn_apility/fn_apility/components/function_apility.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        
+        try:
+            # Get the function parameters:
+            apility_lookup_type = self.get_select_param(kwargs.get("apility_lookup_type"))  # select, values: "Email", "IP Address", "Domain"
+            apility_lookup_value = kwargs.get("apility_lookup_value")  # text
+
+            log = logging.getLogger(__name__)
+            log.info("apility_lookup_type: %s", apility_lookup_type)
+            log.info("apility_lookup_value: %s", apility_lookup_value)
+            # log.info("apility_lookup_type type: %s", type(apility_lookup_value))
```
Comment (mscherfling) [inline fn_apility/fn_apility/components/function_apility.py]: remove commented code if not needed

File: fn_apility/fn_apility/lib/resilient_common.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# (c) Copyright IBM Corp. 2010, 2018. All Rights Reserved.
+
+
+def api_url(query_type):
+    """
+    Method Will return appropriate URL based on query type    
+    """
+    type_url_mapping = {'Email': 'bademail', 'Domain': 'baddomain', 'IP Address': 'v2.0/ip'}
```
Comment (mscherfling) [inline fn_apility/fn_apility/lib/resilient_common.py]: Should you also include 'Email Sender', 'Email Sender Name' and 'Email Recipient' ?

File: fn_apility/fn_apility/util/config.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+"""Generate a default configuration-file section for fn_apility"""
+
+from __future__ import print_function
+
+
+def config_section_data():
+    """Produce the default configuration section for app.config,
+       when called by `resilient-circuits config [-c|-u]`
+    """
+#    config_data = u"""[fn_apility]
```
Comment (mscherfling) [inline fn_apility/fn_apility/util/config.py]: this should be uncommented and match the parameters you referenced in your code (ex: self.options['url'])

File: fn_apility/fn_apility/util/selftest.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+log.setLevel(logging.INFO)
+log.addHandler(logging.StreamHandler())
+
+
+def selftest_function(opts):
+    """
+    Placeholder for selftest function. An example use would be to test package api connectivity.
+    Suggested return values are be unimplemented, success, or failure.
+    """
+    options = opts.get("fn_apility", {})
```
Comment (mscherfling) [inline fn_apility/fn_apility/util/selftest.py]: This should make an api call which confirms that access to this end-point is available for function calls. is there a simple api call which can be coded to ensure connectivity is available?

File: fn_apility/tests/test_fn_apility.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+class TestFnApility:
+    """ Tests for the fn_apility function"""
+
+    def test_function_definition(self):
+        """ Test that the package provides customization_data that defines the function """
+        func = get_function_definition(PACKAGE_NAME, FUNCTION_NAME)
+        assert func is not None
+
+    @pytest.mark.parametrize("apility_lookup_type, apility_lookup_value, expected_results", [
```
Comment (mscherfling) [inline fn_apility/tests/test_fn_apility.py]: Is this being used? The intent is to mock out the results from the endpoint to make sure your code is processing the inputs and outputs correct. We can review if this is not clear.

File: fn_apility/fn_apility/components/function_apility.py (hunk 3 above)
Comment (devsuds) [reply fn_apility/fn_apility/components/function_apility.py]: Ok

File: fn_apility/fn_apility/lib/resilient_common.py (hunk 4 above)
Comment (devsuds) [reply fn_apility/fn_apility/lib/resilient_common.py]: I believe these 3 we discussed for this release/version.

File: fn_apility/fn_apility/util/config.py (hunk 5 above)
Comment (devsuds) [reply fn_apility/fn_apility/util/config.py]: Done!

File: fn_apility/fn_apility/util/selftest.py (hunk 6 above)
Comment (devsuds) [reply fn_apility/fn_apility/util/selftest.py]: So, I have done necessary changes but somehow couldn't run it, plz have a look on that.

File: fn_apility/tests/test_fn_apility.py (hunk 7 above)
Comment (devsuds) [reply fn_apility/tests/test_fn_apility.py]: Done changes as we discussed last time.

File: fn_apility/README.md (hunk 2 above)
Comment (devsuds) [reply fn_apility/README.md]: I have put simple words. Plz check if missing anything.

Comment (mscherfling) [reply fn_apility/README.md]: how about 

> Ability.io is an Internet accessible Threat Intelligence service for ....

Can you describe what 'these lists' mean? Is it ip addresses?

File: fn_apility/fn_apility/lib/resilient_common.py (hunk 4 above)
Comment (mscherfling) [reply fn_apility/fn_apility/lib/resilient_common.py]: Let's include Email Sender and Email Recipient as those will specifically addresses like user@company.co

File: fn_apility/fn_apility/util/selftest.py (hunk 6 above)
Comment (mscherfling) [reply fn_apility/fn_apility/util/selftest.py]: see my comment in slack

**Pull request merged**